internal/integrations/rawtemplate.go

```diff
@@ -0,0 +1,240 @@
+	"go.opentelemetry.io/collector/pipeline"
+)
+
+// RawTemplate implements the Template interface for raw YAML content.
+// Unused components are removed after resolving it, so the variables they contain are not required.
+type RawTemplate struct {
+	source rawYAMLConfig
+}
+
+// NewRawTemplate creates a new RawTemplate from raw YAML content.
+func NewRawTemplate(raw []byte) (*RawTemplate, error) {
+	var source rawYAMLConfig
+	decoder := yaml.NewDecoder(bytes.NewReader(raw))
+	decoder.KnownFields(true)
+	err := decoder.Decode(&source)
+	if err != nil {
+		return nil, fmt.Errorf("invalid integration template format: %w", err)
+	}
+	if err := source.Validate(); err != nil {
+		return nil, fmt.Errorf("integration template validation failed: %w", err)
+	}
+	return &RawTemplate{source: source}, nil
+}
+
+// Resolve resolves the template using a confmap resolver.
+func (t *RawTemplate) Resolve(ctx context.Context, params map[string]any, pipelines []pipeline.ID) (*Config, error) {
+	selectedPipelines := t.source.Pipelines
+	if len(pipelines) > 0 {
+		var err error
+		selectedPipelines, err = selectComponents(t.source.Pipelines, pipelines)
+		if err != nil {
+			return nil, fmt.Errorf("selecting pipelines: %w", err)
+		}
+	}
+	selectedReceivers, err := selectComponents(t.source.Receivers, listReceivers(selectedPipelines))
+	if err != nil {
+		return nil, fmt.Errorf("selecting receivers: %w", err)
+	}
+	selectedProcessors, err := selectComponents(t.source.Processors, listProcessors(selectedPipelines))
+	if err != nil {
+		return nil, fmt.Errorf("selecting processors: %w", err)
+	}
+
+	factory := newConfmapProviderFactory(rawYAMLConfig{
+		Pipelines:  selectedPipelines,
+		Receivers:  selectedReceivers,
+		Processors: selectedProcessors,
+	})
+	resolver, err := newResolver(factory, params)
+	if err != nil {
+		return nil, err
+	}
+	conf, err := resolver.Resolve(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var config Config
+	if err := conf.Unmarshal(&config); err != nil {
+		return nil, fmt.Errorf("failed to generate effective configuration: %w", err)
+	}
+	return &config, config.Validate()
+}
+
+func newResolver(factory confmap.ProviderFactory, variables map[string]any) (*confmap.Resolver, error) {
+	settings := confmap.ResolverSettings{
+		URIs: []string{"config:main"},
+		ProviderFactories: []confmap.ProviderFactory{
+			factory,
+			newVariablesProviderFactory(variables),
+		},
+	}
+	return confmap.NewResolver(settings)
+}
+
+type selectableID interface {
+	comparable
+	fmt.Stringer
+}
+
+func selectComponents[C any, ID selectableID](from map[ID]C, selection []ID) (map[ID]C, error) {
+	if len(selection) == 0 {
+		return nil, nil
+	}
+	selected := make(map[ID]C)
+	for _, id := range selection {
+		component, found := from[id]
+		if !found {
+			return nil, fmt.Errorf("component %q not found", id)
+		}
+		selected[id] = component
+	}
+	return selected, nil
+}
+
+func listReceivers(pipelines map[pipeline.ID]PipelineConfig) []component.ID {
```

Comment:
What is the difference between `listReceivers()` and `listProcessors()`? Can these two functions be merged?

Reply:
They list the receivers and the processors in the map of pipeline configs. Both sets of ids come from different fields.

Refactored to reuse the iteration, sort and compaction.